### ISSUE TITLE
Unable to remove overlays

### ISSUE THREAD

**Issue by TylerVigario** [Kometa-Team/Kometa#1191]
### Version Number

1.18

### What branch are you on?

master

### Describe the Bug

I accidentally added resolution overlays for 1080p and can't remove them.

I have tried:

- `- remove_overlays: true`
- `- reapply_overlays: true`
- `- reset_overlays: plex`

Any ideas on how to remove these overlays?

### Logs

https://gist.github.com/TylerVigario/ab429b878a81fe3b5666bcad6cf1d6cb

**Comment by chazlarson** [comment]
Do you mean you've tried all three of those at once?

**Comment by chazlarson** [comment]
```
- remove_overlays: true
```
should remove the overlays by restoring the backup image and without reapplying any overlays that are defined in the config.

```
- reapply_overlays: true
```
should reapply the same overlays as before, using the backup image as the base, whether or not PMM thinks it has to do so.

```
- reset_overlays: [plex|tmdb]
```
should retrieve a new image from the specified source and apply overlays on that image.

**Comment by G4Zz0L1** [comment]
I've wrote on the discussion section about a similar problem.
With those options theoretically pmm should reset the overlay to the plex default one (the reapply gets ignored, if I understand correctly from the code).
The problem is the fact that on my instance not all overlays get removed.
On some movies/shows/season/episode the overlay remains the same as before (in some cases with a double overlay in different position, like the OP, if I have moved them for testing).
I can't understand why the overlays get doubled in the first place, then there's the "can't delete them" question also.

**Comment by chazlarson** [comment]
Here's one scenario where an overlay could get doubled.  **Note that I am not saying this is what happened in your case**, just offering it as a case where this could happen.

You run an overlay job.  PMM sets the artwork and the "Overlay" label on each item.  Now, you want to put some overlay in a new position, so you mass-remove the "Overlay" label to get PMM to reapply the new overlay.  PMM applies the new overlay, but the base image it pulls from Plex has overlays on it already, so the newly overlaid image has both.  Also, the backup image probably gets overwritten with the previously-overlaid image, so when you try to remove or reapply overlays the base will be the previously-overlaid image.

In your case it could be that the attempt to retrieve a new image from Plex fails, so the art doesn't get reset; I saw this a fair amount when testing this yesterday.  That should be shown in the log.

You  could take a look in the original artwork backup folder to see if some of these things are backed up with overlays already on them.

**Comment by TylerVigario** [comment]
> Here's one scenario where an overlay could get doubled.  **Note that I am not saying this is what happened in your case**, just offering it as a case where this could happen.
> 
> You run an overlay job.  PMM sets the artwork and the "Overlay" label on each item.  Now, you want to put some overlay in a new position, so you mass-remove the "Overlay" label to get PMM to reapply the new overlay.  PMM applies the new overlay, but the base image it pulls from Plex has overlays on it already, so the newly overlaid image has both.  Also, the backup image probably gets overwritten with the previously-overlaid image, so when you try to remove or reapply overlays the base will be the previously-overlaid image.
> 
> In your case it could be that the attempt to retrieve a new image from Plex fails, so the art doesn't get reset; I saw this a fair amount when testing this yesterday.  That should be shown in the log.
> 
> You  could take a look in the original artwork backup folder to see if some of these things are baked up with overlays already on them.

That seems like an issue with PMM.

**Comment by G4Zz0L1** [comment]
The backup folders you mentioned, are those that gets created in the config/overlays folder?
In that case, every file in that folder doesn't have overlays, I've checked the ones that gives this problem (they are always the same, at every run that I try to delete them).
I never touch anything on plex, so I think that this labels problem it's not my case.
I just stop the docker, change the positions in the config (through template variables, I've the pmm default ones), and restart the docker, so the labels shoud be where they were before, am I wrong?
What else could be generating this double overlay?

**Comment by chazlarson** [comment]
Once the overlay label is removed from the item, PMM has no way to tell if the current poster has overlays on it.

If the poster has overlays, and the indicator that this is true is removed, how is PMM to know that the existing image already has overlays and should be discarded?

**Comment by chazlarson** [comment]
The only thing that's going to generate a double-overlay, barring a config that is applying them deliberately, is PMM starting with a base image that already has overlays on it.

I can't explain in a vacuum why this is happening in your specific case, but I tried to break it yesterday and couldn't.

**Comment by chazlarson** [comment]
But back to the original question.

Do you want to remove all overlays, or just the 1080p one and leave some others?

Did you try all three of these at the same time, or one at a time, or what?  In testing yesterday, if I added 
```
- remove_overlays: true
```
under `overlay_path` and left the rest of the overlay config stuff alone, all overlays were removed.

**Comment by TylerVigario** [comment]
> But back to the original question.
> 
> Do you want to remove all overlays, or just the 1080p one and leave some others?
> 
> Did you try all three of these at the same time, or one at a time, or what?  In testing yesterday, if I added 
> ```
> - remove_overlays: true
> ```
> under `overlay_path` and left the rest of the overlay config stuff alone, all overlays were removed.
> 

I tried them separately and at this point removing all of the overlays would work as well. However I do just want the 1080p overlays gone.

In all of your examples I can't understand how it's something I did versus a mishandling of "labels" (as you put) by PMM.

Maybe we need a better method of keeping track of what overlays are applied on any given poster?

For what it's worth this started roughly a month ago when `template_variables` weren't acting right. It erroneously added the 1080p overlays even though I had it set not to. And now I cannot remove them through any official PMM method.

**Comment by chazlarson** [comment]
In my experiments, 
```
- reset_overlays: tmdb
```
Has always grabbed new artwork from tmdb and reapplied appropriate overlays.  if the overlay config no longer includes the 1080p overlay, it won't be applied to the new artwork.

```
- reset_overlays: plex
```
In my experiments has been more hit-or-miss in that sometimes PMM seems to be unable to get new artwork from Plex.

Ordinarily, just removing an individual overlay from the config is sufficient to remove it

For example, going from this:
```
libraries:
  Movies:
    metadata_path:
    - pmm: actor
    - pmm: genre
    overlay_path:
    - pmm: ribbon
    - pmm: ratings
 ```
 to
 ```
libraries:
  Movies:
    metadata_path:
    - pmm: actor
    - pmm: genre
    overlay_path:
    - pmm: ribbon
 ```
And rerunning PMM would remove the ratings overlays without any reset/remove/reapply required.

I didn't intend to say this was your fault.  Just providing an example scenario in response to "I can't understand why the overlays get doubled in the first place".

Chances are this was all initially triggered by some bug or another, but resetting to tmdb artwork has always worked for me.

**Comment by chazlarson** [comment]
There's a brute force way to do just the artwork reset here:
<https://github.com/chazlarson/Media-Scripts>

**Comment by G4Zz0L1** [comment]
I think it's really a pmm thing, I'll explain why I think so.
When I've noticed the first time this "double overlay" behaviour, I've stopped pmm (and every other service that might act inside plex), I've emptied all plex libraries (moving everything out of the libraries and using the beautiful script of bullmoose20 (https://github.com/bullmoose20/Plex-Stuff/blob/main/plex-bloat-fix.py) to clean everything up).
Then I've launched pmm only with a single overlay for a library (like for example the defaults/overlays/resolution) without any template_variable, and the usual stuff on the metadata section.
After the first completed run, I've tried to move it, and it gets doubled only on some elements of the libraries like I was saying before.
So, I had to repeat the emptying part, and trying again, without anything in the metadata or operations, just the single overlay as before, with the same method: first run clean, second run moved.
It gets doubled again.
So, if it works ok for you, this narrows it down to two scenarios:
- I have an error in the config
- pmm bugs something out when it runs on my end

I'll attach my usual config, just to check if everything it's okay, but it's a pretty standard one:

```
libraries:
  Movies:
    report_path: config/Movies_report.yml
    template_variables:
      collection_mode: hide
      use_separator: true
    metadata_path:
      #award
    - pmm: separator_award
    - pmm: bafta
    - pmm: cannes
    - pmm: choice
    - pmm: emmy
    - pmm: golden
    - pmm: oscars
    - pmm: other_award
    - pmm: spirit
    - pmm: sundance
      # both
    - pmm: actor
    - pmm: audio_language
    - pmm: genre
    - pmm: resolution
    - pmm: streaming
    - pmm: studio
    - pmm: studio_anime
    - pmm: subtitle_language
    - pmm: year
      # chart
    - pmm: separator_chart
    - pmm: anilist
    - pmm: basic
    - pmm: flixpatrol
    - pmm: imdb
    - pmm: myanimelist
    - pmm: other_chart
    - pmm: tautulli
    - pmm: tmdb
    - pmm: trakt
      # movie
    - pmm: content_rating_us
    - pmm: country
      template_variables:
        style: color
    - pmm: decade
    - pmm: director
    - pmm: franchise
    - pmm: producer
    - pmm: seasonal
    - pmm: universe
    - pmm: writer
    overlay_path:
    #- remove_overlays: true
    #- reapply_overlay: false
    #- reset_overlays: plex
    - pmm: commonsense
    - pmm: flixpatrol
    - pmm: languages
      template_variables:
        use_it: true
        weight_it: 609
    - pmm: languages
      template_variables:
        use_it: true
        weight_it: 609
        use_subtitles: true
    - pmm: mediastinger
    - pmm: ratings
      template_variables:
        rating1: critic
        rating1_image: imdb
        rating2: audience
        rating2_image: tmdb
        rating3: user
        rating3_image: rt_tomato
    - pmm: resolution
    - pmm: ribbon
    - pmm: video_format
    operations:
      mass_genre_update: tmdb
      mass_content_rating_update: mdb_commonsense
      mass_originally_available_update: tmdb
      mass_audience_rating_update: tmdb
      mass_critic_rating_update: imdb
      mass_user_rating_update: mdb_tomatoes
      mass_imdb_parental_labels: without_none
      mass_collection_mode: hide
      split_duplicates: true
      delete_collections:
        managed: false
        configured: false
        less: 1
      metadata_backup:
        path: config/Movies_Metadata_Backup.yml
        sync_tags: true
        add_blank_entries: false
  Series:
    report_path: config/Series_report.yml
    template_variables:
      collection_mode: hide
      use_separator: true
    metadata_path:
      #award
    - pmm: separator_award
    - pmm: bafta
    - pmm: cannes
    - pmm: choice
    - pmm: emmy
    - pmm: golden
    - pmm: oscars
    - pmm: other_award
    - pmm: spirit
    - pmm: sundance
      # both
    - pmm: actor
    - pmm: audio_language
    - pmm: genre
    - pmm: resolution
    - pmm: streaming
    - pmm: studio
    - pmm: studio_anime
    - pmm: subtitle_language
    - pmm: year
      # chart
    - pmm: separator_chart
    - pmm: anilist
    - pmm: basic
    - pmm: flixpatrol
    - pmm: imdb
    - pmm: myanimelist
    - pmm: other_chart
    - pmm: tautulli
    - pmm: tmdb
    - pmm: trakt
      # show
    - pmm: content_rating_us
    - pmm: country
      template_variables:
        style: color
    - pmm: decade
    - pmm: franchise
    - pmm: network
    overlay_path:
    #- remove_overlays: true
    #- reapply_overlay: false
    #- reset_overlays: plex
    - pmm: commonsense
    - pmm: commonsense
      template_variables:
        overlay_level: season
    - pmm: flixpatrol
    - pmm: languages
      template_variables:
        use_it: true
        weight_it: 609
    - pmm: languages
      template_variables:
        overlay_level: season
        use_it: true
        weight_it: 609
    - pmm: languages
      template_variables:
        overlay_level: episode
        use_it: true
        weight_it: 609
    - pmm: languages
      template_variables:
        use_it: true
        weight_it: 609
        use_subtitles: true
    - pmm: languages
      template_variables:
        overlay_level: season
        use_it: true
        weight_it: 609
        use_subtitles: true
    - pmm: languages
      template_variables:
        overlay_level: episode
        use_it: true
        weight_it: 609
        use_subtitles: true
    - pmm: mediastinger
    - pmm: mediastinger
      template_variables:
        overlay_level: season
    - pmm: mediastinger
      template_variables:
        overlay_level: episode
    - pmm: ratings
      template_variables:
        rating1: critic
        rating1_image: imdb
        rating2: audience
        rating2_image: tmdb
        rating3: user
        rating3_image: rt_tomato
    - pmm: ratings
      template_variables:
        rating1: critic
        rating1_image: imdb
        rating2: audience
        rating2_image: tmdb
        rating3: user
        rating3_image: rt_tomato
        overlay_level: episode
    - pmm: resolution
    - pmm: resolution
      template_variables:
        overlay_level: season
    - pmm: resolution
      template_variables:
        overlay_level: episode
    - pmm: ribbon
    - pmm: ribbon
      template_variables:
        overlay_level: season
    - pmm: ribbon
      template_variables:
        overlay_level: episode
    - pmm: status
    - pmm: video_format
    - pmm: video_format
      template_variables:
        overlay_level: season
    - pmm: video_format
      template_variables:
        overlay_level: episode
    operations:
      mass_genre_update: tmdb
      mass_content_rating_update: mdb_commonsense
      mass_originally_available_update: tmdb
      mass_audience_rating_update: tmdb
      mass_critic_rating_update: imdb
      mass_user_rating_update: mdb_tomatoes
      mass_episode_audience_rating_update: tmdb
      mass_episode_critic_rating_update: imdb
      mass_episode_user_rating_update: tmdb
      mass_imdb_parental_labels: without_none
      mass_collection_mode: hide
      split_duplicates: true
      delete_collections:
        managed: false
        configured: false
        less: 1
      metadata_backup:
        path: config/Series_Metadata_Backup.yml
        sync_tags: true
        add_blank_entries: false
playlist_files:
- file: config/defaults/playlist.yml
settings:
  cache: true
  cache_expiration: 60
  asset_directory: config/assets
  asset_folders: true
  asset_depth: 0
  create_asset_folders: false
  prioritize_assets: false
  dimensional_asset_rename: false
  download_url_assets: false
  show_missing_assets: false
  show_missing_season_assets: false
  show_missing_episode_assets: false
  show_asset_not_needed: false
  sync_mode: sync
  default_collection_order:
  minimum_items: 1
  delete_below_minimum: true
  delete_not_scheduled: false
  run_again_delay: 30
  missing_only_released: true
  show_unmanaged: false
  show_filtered: false
  show_options: true
  show_missing: false
  show_unconfigured: false
  only_filter_missing: false
  save_report: true
  tvdb_language: eng
  ignore_ids:
  ignore_imdb_ids:
  item_refresh_delay: 0
  playlist_sync_to_users:
  playlist_report: true
  custom_repo:
  verify_ssl: true
  check_nightly: true
```

**Comment by chazlarson** [comment]
I can only offer that I've gone through another series of experiments today and I cannot reproduce any unexpected behavior.

I plan to write this up with screenshots as a help article, and maybe someone sees differences to how they are using the feature.

**Comment by G4Zz0L1** [comment]
So, I've no choice to use this feature without some errors?   xD
Maybe it's the size of my libraries, but I can't be the only one with a lot of stuff on plex, I've got like 5k movies and 1k series, a normal run with the config above lasts about 1.5-2 day.
That's the reason why I've been experimenting with the `reapply_overlay` flag, I don't need all the overlays recreated on every single item, but only if something changes on that element.
But that doesn't work either on my part, so I've given up and removed all overlays altogether.
It sucks that this can't be solved easily because it's a wonderful feature.

**Comment by TylerVigario** [comment]
> In my experiments, 
> ```
> - reset_overlays: tmdb
> ```
> Has always grabbed new artwork from tmdb and reapplied appropriate overlays.  if the overlay config no longer includes the 1080p overlay, it won't be applied to the new artwork.
> 
> ```
> - reset_overlays: plex
> ```
> In my experiments has been more hit-or-miss in that sometimes PMM seems to be unable to get new artwork from Plex.
> 

I'll give this a shot when I get home

**Comment by G4Zz0L1** [comment]
I've relaunched pmm with the overlays and I'm checking the labels.
It seems that the `Overlay` labels doesn't get applied on some items, even if the log doesn't show errors or anything strange as it applies the overlays.
So, that's the reason for the doubled overlays.
Now I've to understand why it happens, only on some items.

EDIT: like a lot of items, it's almost 30-40% with overlays without labels, according to the plex filters.

**Comment by chazlarson** [comment]
Odd; that's not something I've ever seen.

**Comment by G4Zz0L1** [comment]
Don't know what to say, the docker should be the same as the rest, and the config is the one posted before.
For the series library it's even worse, we're like 60-70% with overlays and no labels.

**Comment by chazlarson** [comment]
More experiments today, and I *think* I've found a problem and a solution for this edge case.  I could only make it happen by forcing it, so it's likely that the initial situation arises due to some bug or other, and I can't reproduce or explain the "only applies overlay label to some of the items" scenario.

Basically, the "problem" is that the `reset_overlays: tmdb` operation doesn't get applied to items without the "Overlay" label, so here was what I did to recover in my test library:

1. empty the backup images directory for the library, if there's anything in there.
2. manually apply "Overlay" label to everything in the library
3. run with `reset_overlays: tmdb`.  Now everything got a new poster with overlays applied and the backup images are all non-overlaid.
4. From here, if you want to remove all overlays, do a run with `remove_overlays: true`, then remove the overlay section from your config.
5. if you want to adjust the overlays [like getting rid of the 1080p overlay], edit the config as usual, remove the `remove_overlays: true`, and run PMM.

I'm planning to write this up with all the screenshots and configs I just saved and will post that here and in the discord.

Another option outside PMM would be to remove all the "Overlay" labels from everything and run the `reset_posters_tmdb.py` script from the repo I linked above.

**Comment by G4Zz0L1** [comment]
I don't think the situation is caused by an external factor, it's the same even if I manually reset the library with the steps listed above:
- moving everything out of the library
- clearing pmm backup folders
- clearing plex metadata with the script
- moving the files back in the library, as a clean start

I've noticed that even in your script's README, you mention some problems with the labels, I think it's the same reason, but unknown cause.

![image](https://user-images.githubusercontent.com/13220636/203877627-54705c6c-0ece-47a2-ae4c-dec5cc37707e.png)

At this point I think the problem lies on plex part, not pmm, maybe for some internal reason the labels aren't reliable, but I don't know how to test it for being sure.

**Comment by TylerVigario** [comment]
> 2. manually apply "Overlay" label to everything in the library

~~Any easy way to do this? I have an expansive library.~~ This did the trick!

Also, would it be more reliable to keep track of these labels internally in a database?

**Comment by chazlarson** [comment]
I don't think keeping track externally helps very much.  It's still a flag keeping track of whether a thing is in place or not, and presumably having the flag on the item in Plex's database is more reliable than having it in an external database that can be deleted at any time.

Leaving aside this mysterious thing where sometimes G4Zz0L1's labels don't get updated consistently.

**Comment by TylerVigario** [comment]
> I don't think keeping track externally helps very much. It's still a flag keeping track of whether a thing is in place or not, and presumably having the flag on the item in Plex's database is more reliable than having it in an external database that can be deleted at any time.
> 
> Leaving aside this mysterious thing where sometimes G4Zz0L1's labels don't get updated consistently.

I respectfully disagree. Right now, PMM relies on a field within Plex (also a database) which is easy to manipulate within the Plex web interface. This also opens up the opportunity that Plex doesn't properly set the flag or remove it for whatever reason, and there isn't much we can do about that.

EDIT: Rephrased

**Comment by chazlarson** [comment]
> I respectfully disagree. Right now you rely on a field within Plex (also a database) which is easy to manipulate within the Plex web interface. This also opens up the opportunity that Plex doesn't properly set the flag or remove it for whatever reason, and there isn't much you can do about that.

Any solution has tradeoffs, it's true.  Ultimately it's up to the author to make any changes he sees fit.

**Comment by TylerVigario** [comment]
> Any solution has tradeoffs, it's true. Ultimately it's up to the author to make any changes he sees fit.

Agreed. Thanks again for all your help!

**Comment by G4Zz0L1** [comment]
> Leaving aside this mysterious thing where sometimes G4Zz0L1's labels don't get updated consistently.

@chazlarson It's not mysterious, it's happened to you too, as far as I read on your repo.
There isn't something I can test, to check if the problem with these labels it's on plex side or on pmm?

**Comment by chazlarson** [comment]
I’ve never seen the case you describe where PMM, while setting overlays, only sets the overlay label on *some* of the items.  I have seen a limited issue on a fixed set of items in a library when attempting to remove labels.  In my case, it was always the same few movies.  I haven’t seen a general problem adding or removing labels.  The reason I described it as mysterious is that it’s not a common problem and apparently doesn’t have a repeatable recipe.  It could be a database issue, or a PlexAPI issue, or a Plex issue.  It’s mysterious in that it has an unknown cause and can’t be reproduced at will.

**Comment by G4Zz0L1** [comment]
So, in short, there is a feature of pmm I can't use, because of this and the fact it can't be narrowed down to be solved   xD
As you've writed on you repo, the percentage of failure to edit labels (add or remove it's the same, for that matter) is roughly the same here, and on the same fixed set of items every run as you say, so I thought it was the same reason.
Maybe the devs can shed some light here, but if it's as you say it's not a thing that can be solved here on pmm, as far as I understand.
I'll try to wrap my head around this and give up, thanks for the help.

**Comment by G4Zz0L1** [comment]
I've tried to search this on my own, even if I'm not great with python.
Can this be the cause of this issue? https://github.com/pkkid/python-plexapi/issues/988
I've searched around the code, if I'm not mistaken the only point where the labels are added is this one:
https://github.com/meisnate12/Plex-Meta-Manager/blob/nightly/modules/plex.py#L749
I don't see a `reload` like in the other issue, and if it's the same problem, maybe a fix on the PMM side is required to work, right?
Or maybe I overreached and it's not the same, tell me if I'm wrong.
